### PR TITLE
Unify Index._dir_* with Series implementation

### PR DESCRIPTION
**Pull request by jbrockmendel**
Have `Index` use the same implementation of `_dir_additions` and `_dir_deletions` that `Series` currently uses. Move that implementation into a shared base class. @jreback suggested that shared base class `DirNamesMixin` be moved from `core.base` to `core.accessor`.

<s>De-duplicate `_dir_additions` and `_dir_deletions` code.  `Index` currently gets its implementations of these two methods from `strings.StringMethodMixin`, but the implementation is just a special case of the implementation in `Series`.  That implementation is essentially The General Case, so this PR moves it up to a base class `DirNamesMixin`.  (I'm open to putting this elsewhere in the hierarchy)</s>

This PR removes `StringMethodMixin`, discussed in #17042.

<s>For `DatetimeIndex`, `TimeDeltaIndex`, and `PeriodIndex`, this adds a property `dt` that just returns `self`.  This way a user can access e.g. `item.year` symmetrically without having to check whether `item` is a `Series/Index`.

Holding off on writing tests until a consensus is reached.</s>

 - [ ] closes #xxxx
 - [ ] tests added / passed
 - [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
ok your re-factor to string out the accessors is ok. but pls put in a new ``core/accessor.py`` which can be imported into base (IOW it has no other deps).

also remove the .dt/.cat/.str accessors from the Index objects. pls make an issue about this instead; I think its ok, but would like some discussion.

In general try to do 1 change per PR to avoid intermixing things.

I think existing testing should be sufficient for the accessor re-factor (but ok to add if something is not covered).

ping on green.

**Comment by jbrockmendel** [comment]
> ok your re-factor to string out the accessors is ok. but pls put in a new core/accessor.py which can be imported into base (IOW it has no other deps).

I don't understand this sentence.  What is it that you want to keep and what is it that you want in core/accessor.py?  There is a small amount of overlap between this and #17042, but I tried to keep that to a minimum.

> also remove the .dt/.cat/.str accessors from the Index objects. pls make an issue about this instead; I think its ok, but would like some discussion.

OK

**Comment by jreback** [comment]
> ok your re-factor to string out the accessors is ok. but pls put in a new core/accessor.py which can be imported into base (IOW it has no other deps).

- move the changes that you made in ``core/base.py`` to a new ``core/accessor.py``.
- remove the ``.dt`` etc from ``Index`` for now. (the changes you made to Series are fine).

File: pandas/core/base.py [hunk 1]
```diff
@@ -72,7 +72,35 @@ def __repr__(self):
         return str(self)
 
 
-class PandasObject(StringMixin):
+class DirNamesMixin(object):
```
Comment (jreback) [inline pandas/core/base.py]: move this class to ``core/accessor.py``

Comment (jbrockmendel) [inline pandas/core/base.py]: Got it, thanks.

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -151,6 +151,12 @@ class Index(IndexOpsMixin, StringAccessorMixin, PandasObject):
 
     _engine_type = libindex.ObjectEngine
 
+    _accessors = frozenset(['dt', 'str', 'cat'])
+
+    # String Methods
+    str = base.AccessorProperty(strings.StringMethods,
```
Comment (jreback) [inline pandas/core/indexes/base.py]: rm this

File: pandas/core/indexes/datetimelike.py [hunk 3]
```diff
@@ -125,6 +125,15 @@ def ceil(self, freq):
 class DatetimeIndexOpsMixin(object):
     """ common ops mixin to support a unified inteface datetimelike Index """
 
+    @property
+    def dt(self):
+        """
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: rm this

File: pandas/core/series.py [hunk 4]
```diff
@@ -2934,18 +2933,9 @@ def _make_cat_accessor(self):
...
-        rv = set()
-        for accessor in self._accessors:
-            try:
-                getattr(self, accessor)
-                rv.add(accessor)
-            except AttributeError:
-                pass
-        return rv
+    # String Methods
+    str = base.AccessorProperty(strings.StringMethods,
```
Comment (jreback) [inline pandas/core/series.py]: this is ok

File: pandas/core/strings.py [hunk 5]
```diff
@@ -1910,30 +1907,18 @@ def _make_str_accessor(self):
...
                 message = ("Can only use .str accessor with string values "
                            "(i.e. inferred_type is 'string', 'unicode' or "
                            "'mixed')")
                 raise AttributeError(message)
-            if self.nlevels > 1:
+            if data.nlevels > 1:
                 message = ("Can only use .str accessor with Index, not "
                            "MultiIndex")
                 raise AttributeError(message)
-        return StringMethods(self)
```
Comment (jreback) [inline pandas/core/strings.py]: this is ok

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jbrockmendel) [inline pandas/core/indexes/base.py]: Then should I revert deletion of `StringAccessorMixin` and mix that back in to `Index`?

**Comment by jreback** [comment]
closing, this is not the impl we want and is confusing.

**Comment by jreback** [comment]
sorry this one was changed to be reasonable. still need to review.

File: pandas/core/accessor.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
```
Comment (jreback) [inline pandas/core/accessor.py]: can you add a comment about what this module contains

File: pandas/core/accessor.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
```
Comment (jreback) [inline pandas/core/accessor.py]: remove the first line, we don't normally do this

**Review by jreback — APPROVED** [review]
tiny comments. ping on green.

**Comment by jorisvandenbossche** [comment]
@jbrockmendel I wanted to review this, but before doing that, some questions/requests:

* can you give some updated explanation of what the PR is now doing? As I thought the scope has changed a bit during its lifetime (or not?) 
* Can you give the PR a more descriptive title?
* Is this only internal refactorings in the end? As I don't see a whatsnew note or changes to tests (but if this actually changes behaviour, that should be tested/noted)

**Comment by jbrockmendel** [comment]
@jorisvandenbossche 

> can you give some updated explanation of what the PR is now doing?

You're absolutely right that this has meandered over its lifetime.  It's now pretty limited in scope.

TLDR: Have `Index` use the same implementation of `_dir_additions` and `_dir_deletions` that `Series` currently uses.  Move that implementation into a shared base class.  @jreback suggested that shared base class `DirNamesMixin` be moved from `core.base` to `core.accessor`.

This lets us get rid of `core.strings.StringAccessorMixin` which is currently used about 1.33 times.

> if appropriate, you can also update the top post

Will do.

> Can you give the PR a more descriptive title?

Will do.

> Is this only internal refactorings in the end?

Very nearly.  It is technically changing the behavior of `Index._dir_additions` and `_dir_deletions`, adding an `_accessors` attribute (to match `Series`).  In conjunction with #17134 (pending your signing off on that I'll make a PR) it will handle the `dt` attribute correctly.

File: pandas/core/accessor.py (hunk 7 above)
Comment (jbrockmendel) [inline pandas/core/accessor.py]: Just pushed a commit that does this and adds a comment in the docstring as requested below.  The docstring isn't super-useful at this point because there isn't much here.  After this is merged I'll rebase #17042 and accessor.py will have a coherent theme.

File: pandas/core/indexes/base.py [hunk 8]
```diff
@@ -151,6 +151,11 @@ class Index(IndexOpsMixin, StringAccessorMixin, PandasObject):
 
     _engine_type = libindex.ObjectEngine
 
+    _accessors = frozenset(['dt', 'str', 'cat'])
+
```
Comment (jreback) [inline pandas/core/indexes/base.py]: there are no tests for the non string accessors in Index

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py [hunk 9]
```diff
@@ -151,6 +151,11 @@ class Index(IndexOpsMixin, StringAccessorMixin, PandasObject):
 
     _engine_type = libindex.ObjectEngine
 
+    _accessors = frozenset(['dt', 'str', 'cat'])
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/base.py]: maybe leave out 'dt' and 'cat' until they are actually added as accessor? (in a next PR I suppose)

File: pandas/core/generic.py [hunk 10]
```diff
@@ -192,8 +192,9 @@ def __unicode__(self):
 
     def _dir_additions(self):
         """ add the string-like attributes from the info_axis """
-        return set([c for c in self._info_axis
-                    if isinstance(c, string_types) and isidentifier(c)])
+        additions = set([c for c in self._info_axis
+                         if isinstance(c, string_types) and isidentifier(c)])
+        return PandasObject._dir_additions(self).union(additions)
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: Is there a reason not to use `super` ?

File: pandas/core/accessor.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+that can be mixed into or pinned onto other pandas classes.
+
+"""
+
+class DirNamesMixin(object):
+    _accessors = frozenset([])
+
+    def _dir_deletions(self):
+        """ delete unwanted __dir__ for this object """
+        return self._accessors
```
Comment (jorisvandenbossche) [inline pandas/core/accessor.py]: can you put a comment why accessors are deleted? (as only the relevant ones are added back afterwards?)

**Review by jorisvandenbossche — COMMENTED** [review]
This looks like a good improvement! (more logical)

File: pandas/core/indexes/base.py (hunk 9 above)
Comment (jorisvandenbossche) [inline pandas/core/indexes/base.py]: ah, but they are not yet really added as accessor (so no user visible change), only 'removed' from `__dir__` (but not really removed since they are not there yet), correct? 
In that case not that important.

I suppose we already have tests to check that `'str'` is only present on the correct index types?

**Comment by jorisvandenbossche** [comment]
>  @jreback suggested that shared base class DirNamesMixin be moved from core.base to core.accessor.

I still feel this seems more appropriate to put in base.py, as it is about adding `__dir__` to the base objects, and is only used in base.py

File: pandas/core/generic.py (hunk 10 above)
Comment (jbrockmendel) [inline pandas/core/generic.py]: Not really.  I developed the habit working in `statsmodels` where inheritance chains can get pretty messy.  Happy to change it.

File: pandas/core/indexes/base.py (hunk 9 above)
Comment (jbrockmendel) [inline pandas/core/indexes/base.py]: > I suppose we already have tests to check that 'str' is only present on the correct index types?

I'm not sure off the top, but I wrote some tests for #17204 that can be adapted pretty readily.

File: pandas/core/generic.py (hunk 10 above)
Comment (jreback) [inline pandas/core/generic.py]: super is the idiom. Calling specifically named methods usually indicates something odd in the inheritance chain.

File: pandas/core/indexes/base.py (hunk 9 above)
Comment (jreback) [inline pandas/core/indexes/base.py]: pls remove dt/cat, leave for another PR. str is covered I believe.

**Comment by jreback** [comment]
> I still feel this seems more appropriate to put in base.py, as it is about adding __dir__ to the base objects, and is only used in base.py

This is just as clean and removes some boilerplate from there. So happy where it is.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
If this is good to go, I can rebase #17042 and tie that off, along with #17134.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!
